onnxruntime/core/providers/cpu/tensor/scatter.cc

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//https://github.com/onnx/onnx/blob/master/docs/Operators.md#Scatter
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/providers/common.h"
+
+namespace onnxruntime {
+
+class Scatter final : public OpKernel {
+ public:
+  Scatter(const OpKernelInfo& info) : OpKernel(info) {
+    ORT_ENFORCE(info.GetAttr<int64_t>("axis", &axis_).IsOK(),
+                "Missing/Invalid 'axis' attribute value");
+  }
+  ~Scatter() = default;
+  Status Compute(OpKernelContext* context) const override;
+
+ private:
+  int64_t axis_;
+};
+
+ONNX_CPU_OPERATOR_KERNEL(
+    Scatter,
+    9,
+    KernelDefBuilder()
+        .MayInplace(0, 0)
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+        .TypeConstraint("Tind", std::vector<MLDataType>{DataTypeImpl::GetTensorType<int32_t>(), DataTypeImpl::GetTensorType<int64_t>()}),
+    Scatter);
+
+template <class Tin>
+Status CopyScatterData(const Tensor* data_input, const Tensor* indecies_input, const Tensor* updates_input,
+                       const int64_t axis, Tensor* data_output) {
+  const TensorShape& input_data_shape = data_input->Shape();
+  const Tin* indices_data = indecies_input->template Data<Tin>();
+  const auto num_indices = indecies_input->Shape().Size();
+  for (int64_t i = 0; i < num_indices; ++i) {
+    Tin idx = indices_data[i];
+    if (idx < 0 || idx >= input_data_shape[axis]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "indices element out of data bounds, idx=", idx,
+                             " data_dim=", input_data_shape[axis]);
+    }
+  }
+
+  const auto input_elements = input_data_shape.Size();
+  const auto element_bytes = data_input->DataType()->Size();
+  const auto total_input_bytes = input_elements * element_bytes;
+
+  const uint8_t* src_base = reinterpret_cast<const uint8_t*>(data_input->DataRaw());
+  uint8_t* dst_base = reinterpret_cast<uint8_t*>(data_output->MutableDataRaw());
+  const bool is_string_type = data_input->DataType() == DataTypeImpl::GetType<std::string>();
+
+  // We allow runtime to re-use input for output. If input/output Tensor* are the same
+  // we do not copy
+  if (data_input != data_output) {
```

Comment:
>data_input != data_output [](start = 6, length = 25)

Will these be the same Tensor*, or should this be src_base != dst_base?

Reply:
I think comparing bases would always work.

---
In reply to: [261004436](https://github.com/Microsoft/onnxruntime/pull/523#discussion_r261004436) [](ancestors = 261004436)

Reply:
Yes, I think it should be comparing bases